ddtrace/writer.py

```diff
@@ -105,51 +106,99 @@ def _on_shutdown(self):
-                try:
-                    traces = self._apply_filters(traces)
-                except Exception as err:
-                    log.error('error while filtering traces:{0}'.format(err))
-            if traces:
-                # If we have data, let's try to send it.
-                try:
-                    traces_response = self.api.send_traces(traces)
-                except Exception as err:
-                    log.error('cannot send spans to {1}:{2}: {0}'.format(err, self.api.hostname, self.api.port))
-
-            if self._trace_queue.closed() and self._trace_queue.size() == 0:
-                # no traces and the queue is closed. our work is done
+            # DEV: Returns `None` or a Trace
+            trace = self._trace_queue.get()
+
+            # Filter the trace
+            # DEV: We set `trace` to `None` on an error because we want to continue the loop
+            #      since this loop needs to execute in order to flush a payload if enough time
+            #      has passed since the last flush.
+            try:
+                # DEV: Returns the trace or `None` if it was filtered
+                trace = self._filter_trace(trace)
+            except Exception as err:
+                log.error('Error raised while trying to filter trace %r: %s', trace, err)
+                trace = None
+
+            if trace:
+                log.debug('Popped trace with %s spans', len(trace))
+                payload.add_trace(trace)
+
+            now = time.time()
+            # If the payload is large enough or enough time has passed, then flush
+            if payload.full or now >= next_flush:
+                if not payload.empty:
+                    log.debug('Attempting to flush %r', payload)
+                    self._flush(payload)
+
+                    # Get a fresh payload
+                    # DEV: Payloads are not reusable
+                    payload = self.api.new_payload()
+                next_flush = now + 1
+
+            # Wait up until the next flush for another trace to be added
+            self._trace_queue.wait(max(0, next_flush - now))
+
+            # no traces and the queue is closed. our work is done
+            if self._trace_queue.closed() and self._trace_queue.qsize() == 0:
+                if not payload.empty:
+                    self._flush(payload)
+                    payload.reset()
+                log.debug('Trace queue closed and empty, exiting')
                 return
 
-            if self._priority_sampler and traces_response:
-                result_traces_json = traces_response.get_json()
-                if result_traces_json and 'rate_by_service' in result_traces_json:
-                    self._priority_sampler.set_sample_rate_by_service(result_traces_json['rate_by_service'])
+    def _flush(self, payload):
+        # Nothing to do, return early
+        if payload.empty:
+            log.debug('Payload empty, not flushing %r', payload)
+            return
```

Comment:
On L161 and L176 there are already emptiness checks.
I feel like it should be decided where the check should be done and done only once and for all.